libcudacxx/include/cuda/std/__bit/clz.h

```diff
@@ -22,131 +22,97 @@
-}
-
-#if !_CCCL_COMPILER(MSVC)
-
-_LIBCUDACXX_HIDE_FROM_ABI constexpr int __constexpr_clz(uint32_t __x) noexcept
-{
-#  if defined(__CUDA_ARCH__)
-  return __binary_clz32(static_cast<uint64_t>(__x), 0); // no device constexpr builtins
-#  else
-  return __builtin_clz(__x);
-#  endif
-}
-
-_LIBCUDACXX_HIDE_FROM_ABI constexpr int __constexpr_clz(uint64_t __x) noexcept
-{
-#  if defined(__CUDA_ARCH__)
-  return __binary_clz64(__x); // no device constexpr builtins
-#  else
-  return __builtin_clzll(__x);
-#  endif
-}
-
-_LIBCUDACXX_HIDE_FROM_ABI constexpr int __cccl_clz(uint32_t __x) noexcept
-{
-#  if _CCCL_STD_VER >= 2014
-  if (!__cccl_default_is_constant_evaluated())
+#if defined(_CCCL_BUILTIN_CLZ)
+  return _CCCL_BUILTIN_CLZ(__x);
+#else
+  constexpr auto __digits = numeric_limits<uint32_t>::digits;
+  if (__x == 0)
   {
-    NV_IF_ELSE_TARGET(NV_IS_DEVICE, (return __clz(__x);), (return __builtin_clz(__x);))
+    return __digits;
   }
-#  endif
-  return __constexpr_clz(__x);
-}
-
-_LIBCUDACXX_HIDE_FROM_ABI constexpr int __cccl_clz(uint64_t __x) noexcept
-{
-#  if _CCCL_STD_VER >= 2014
-  if (!__cccl_default_is_constant_evaluated())
+  unsigned __res = 0;
+  for (unsigned __i = __digits / 2; __i >= 1; __i /= 2)
   {
-    NV_IF_ELSE_TARGET(NV_IS_DEVICE, (return __clzll(__x);), (return __builtin_clzll(__x);))
+    const auto __mark = (~uint32_t{0} >> (__digits - __i)) << __i;
+    if (__x & __mark)
+    {
+      __x >>= __i;
+      __res |= __i;
+    }
   }
-#  endif
-  return __constexpr_clz(__x);
+  return __digits - 1 - __res;
+#endif // defined(_CCCL_BUILTIN_CLZ)
 }
 
-#else // _CCCL_COMPILER(MSVC)
-
-// Precondition:  __x != 0
-_LIBCUDACXX_HIDE_FROM_ABI constexpr int __cccl_clz(uint32_t __x)
+template <typename _Tp>
+_LIBCUDACXX_HIDE_FROM_ABI constexpr int __constexpr_clz(_Tp __x) noexcept
 {
-#  if !defined(__CUDA_ARCH__)
-  if (!__cccl_default_is_constant_evaluated())
+  static_assert(is_same_v<_Tp, uint32_t> || is_same_v<_Tp, uint64_t>);
+  if constexpr (sizeof(_Tp) == sizeof(uint32_t))
   {
-    unsigned long __where = 0;
-    if (_BitScanReverse(&__where, __x))
-    {
-      return static_cast<int>(31 - __where);
-    }
-    return 32; // Undefined Behavior.
+    return __constexpr_clz_32bit(__x);
+  }
+  else
+  {
+#if defined(_CCCL_BUILTIN_CLZLL)
+    return _CCCL_BUILTIN_CLZLL(__x);
+#else
+    auto __higher = __constexpr_clz_32bit(static_cast<uint32_t>(__x >> 32));
+    return __higher != numeric_limits<uint32_t>::digits
+           ? __higher
+           : numeric_limits<uint32_t>::digits + __constexpr_clz_32bit(static_cast<uint32_t>(__x));
+#endif // defined(_CCCL_BUILTIN_CLZLL)
   }
-#  endif // _CCCL_BUILTIN_IS_CONSTANT_EVALUATED && !defined(__CUDA_ARCH__)
-
-  return __binary_clz32(static_cast<uint64_t>(__x), 0);
 }
 
-_LIBCUDACXX_HIDE_FROM_ABI constexpr int __cccl_clz(uint64_t __x)
+template <typename _Tp>
+_LIBCUDACXX_HIDE_FROM_ABI int __host_runtime_clz(_Tp __x) noexcept
```

Comment:
This function is implicitly marked as `__host__ __device__` through `_LIBCUDACXX_HIDE_FROM_ABI` 

I believe this should rather be `_CCCL_HIDE_FROM_ABI` and also #ifdefed out for NVRTC